### PR TITLE
fix: refactoring check control lost logic.

### PR DESCRIPTION
**Pull request by rodrigoleonello**

### CODE REVIEW

File: tasks/DroneFlightControlTask.cpp [hunk 1]
```diff
@@ -95,15 +95,9 @@ DroneFlightControlTask::runtimeStatesTransition(Telemetry::SDKInfo control_statu
     }
 }
 
-void DroneFlightControlTask::applyTransition(
-    DroneFlightControlTask::States const& next_state)
+bool DroneFlightControlTask::transitionToControlling(
```
Comment (jhonasiv) [inline tasks/DroneFlightControlTask.cpp]: I guess returning the State would be cleaner, no?

File: tasks/DroneFlightControlTask.cpp [hunk 2]
```diff
@@ -146,27 +139,38 @@ void DroneFlightControlTask::updateHook()
...
+            }
+            else
+            {
+                state(CONTROL_LOST);
+            }
+        }
+        else
+        {
+            state(new_state);
+        }
```
Comment (jhonasiv) [inline tasks/DroneFlightControlTask.cpp]: Changing transitionToControlling to return a State would dramatically improve this. That way you could do:
```
auto resulting_state = transitionToControlling(new_state)
state(resulting_state)
```

File: tasks/DroneFlightControlTask.cpp [hunk 3]
```diff
@@ -604,11 +608,21 @@ bool DroneFlightControlTask::teardownSubscription(const int pkgIndex)
...
+    auto dji_flight_status =
+        mVehicle->subscribe->getValue<Telemetry::TOPIC_STATUS_FLIGHT>();
+    mStatus.control_device = static_cast<ControlDevice>(control_device.deviceStatus);
+    mStatus.device_flight_status =
+        static_cast<DeviceFlightStatus>(control_device.flightStatus);
+    _status.write(mStatus);
+    drone_control::FlightStatus flight_status =
+        static_cast<drone_control::FlightStatus>(dji_flight_status);
+    _flight_status.write(flight_status);
+
```
Comment (jhonasiv) [inline tasks/DroneFlightControlTask.cpp]: Turn this into another function and call it both here and at the `updateHook`. The way it is rn, we would only output telemetry data when we were checking if the sdk could take control.

File: tasks/DroneFlightControlTask.cpp (hunk 2 above)
Comment (rodrigoleonello) [reply tasks/DroneFlightControlTask.cpp]: ~Could be. The question is: the state transition should be done completely in the updateHook() or in different parts of the code?~

File: tasks/DroneFlightControlTask.cpp (hunk 3 above)
Comment (rodrigoleonello) [reply tasks/DroneFlightControlTask.cpp]: definitely. Thanks

File: tasks/DroneFlightControlTask.cpp (hunk 2 above)
Comment (rodrigoleonello) [reply tasks/DroneFlightControlTask.cpp]: Oh, I removed the others states transitions. In that case, that is the only one. Nice

Comment (jhonasiv) [reply tasks/DroneFlightControlTask.cpp]: It's better to centralize it on the `updateHook`, but this wouldnt change that. You would only output the resulting state from transitionToControlling, while the transition would still happen in the `updateHook`.

Comment (doudou) [reply tasks/DroneFlightControlTask.cpp]: I'm actually not 100% on this side ... I like to boolean as well in this case, as we see all state transitions explcitly in the updateHook. Otherwise we would have "whatever state is returned by transitionToControlling"

Obviously, this works in this case only because there are only two possible outcomes to transitionToControlling

File: tasks/DroneFlightControlTask.cpp [hunk 4]
```diff
@@ -146,27 +139,38 @@ void DroneFlightControlTask::updateHook()
     States new_state = runtimeStatesTransition(control_device);
     if (state() != new_state)
     {
-        applyTransition(new_state);
+        if (new_state != CONTROLLING)
```
Comment (doudou) [inline tasks/DroneFlightControlTask.cpp]: Inverted logic. Should be `if (new_state == CONTROLLING)`

File: tasks/DroneFlightControlTask.cpp [hunk 5]
```diff
@@ -112,20 +106,19 @@ void DroneFlightControlTask::applyTransition(
...
      * feedback that this was successfull might be late.
      */
-    while (now < deadline && !canTakeControl(control_device))
+    while (now < deadline)
     {
         auto timeout = (deadline - now).toSeconds();
         mAuthorityStatus = mVehicle->obtainCtrlAuthority(timeout);
         ACK::getErrorCodeMessage(mAuthorityStatus, __func__);
         control_device = mVehicle->subscribe->getValue<Telemetry::TOPIC_CONTROL_DEVICE>();
+        if (!canTakeControl(control_device))
```
Comment (jhonasiv) [inline tasks/DroneFlightControlTask.cpp]: This shouldn't be negated. We return true if we took control (i.e. `canTakeControl(control_device)` returns true), and return false if couldnt take control by the deadline.

File: tasks/DroneFlightControlTask.cpp (hunk 2 above)
Comment (jhonasiv) [reply tasks/DroneFlightControlTask.cpp]: I can see the point of leaving it, as it makes the logic explicit. I think this is too wordy, so that's a tradeoff. I guess clarity trumps cleanliness.

File: tasks/DroneFlightControlTask.cpp (hunk 5 above)
Comment (rodrigoleonello) [reply tasks/DroneFlightControlTask.cpp]: Now that the while logic is different, let's keep that negative.

File: tasks/DroneFlightControlTask.cpp [hunk 6]
```diff
@@ -112,20 +106,19 @@ void DroneFlightControlTask::applyTransition(
...
+    while (!canTakeControl(control_device))
     {
         auto timeout = (deadline - now).toSeconds();
         mAuthorityStatus = mVehicle->obtainCtrlAuthority(timeout);
         ACK::getErrorCodeMessage(mAuthorityStatus, __func__);
         control_device = mVehicle->subscribe->getValue<Telemetry::TOPIC_CONTROL_DEVICE>();
+        if (now < deadline)
+        {
+            return CONTROL_LOST;
+        }
```
Comment (jhonasiv) [inline tasks/DroneFlightControlTask.cpp]: Move it to the start or to the end of the loop. To me, the end is prefer as we would try to take control at least once.

File: tasks/DroneFlightControlTask.cpp [hunk 7]
```diff
@@ -608,15 +600,8 @@ bool DroneFlightControlTask::teardownSubscription(const int pkgIndex)
     return true;
 }
 
-bool DroneFlightControlTask::canTakeControl(Telemetry::SDKInfo const& control_device)
+void DroneFlightControlTask::checkFlightStatus()
```
Comment (jhonasiv) [inline tasks/DroneFlightControlTask.cpp]: rename it to `writeStatus()`?

Comment (rodrigoleonello) [reply tasks/DroneFlightControlTask.cpp]: done

**Comment by jhonasiv** [comment]
Merging this as it passed CI (https://buildbot.tidewise.io/#/builders/2047/builds/3)
